DFGOSRAvailabilityAnalysis phase should only assert liveness of arguments during validation pass

danlliu opened the pull request:
#### 92d176326e5c5132fa244ff9790807f9e8daddf9
<pre>
DFGOSRAvailabilityAnalysis phase should only assert liveness of arguments during validation pass
<a href="https://bugs.webkit.org/show_bug.cgi?id=301773">https://bugs.webkit.org/show_bug.cgi?id=301773</a>
<a href="https://rdar.apple.com/158595919">rdar://158595919</a>

Reviewed by Dan Hecht.

It is incorrect to assert liveness while we are still converging to fixpoint;
we should only assert once we&apos;ve converged to fixpoint. This fix ensures that
we only check that arguments are live to calls when we have reached fixpoint.

* Source/JavaScriptCore/dfg/DFGOSRAvailabilityAnalysisPhase.cpp:
(JSC::DFG::OSRAvailabilityAnalysisPhase::run):
(JSC::DFG::LocalOSRAvailabilityCalculator::executeNode):
(JSC::DFG::LocalOSRAvailabilityCalculator::validateNode):
* Source/JavaScriptCore/dfg/DFGOSRAvailabilityAnalysisPhase.h:

Canonical link: <a href="https://commits.webkit.org/302616@main">https://commits.webkit.org/302616@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cb24326f00625353884a1b656f136b3d79668720

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129715 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1976 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40571 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137104 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/81179 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/650663b4-3962-4071-a7aa-738312d7d475) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1925 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1865 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/98813 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/81179 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/1e3bf104-b9b2-4522-88ef-ecb6fac90bbd) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132662 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/1468 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/116174 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79491 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8d1d0041-3f32-4a77-a957-8d04a66a60fe) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/1386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34307 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80376 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/121707 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/109863 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34809 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139586 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/128167 "Built successfully and passed tests") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/1770 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/1659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/107323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1815 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112523 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107195 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27283 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/164/builds/1437 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/31020 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54520 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1843 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/65212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/161181 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/1657 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40195 "Passed tests") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/1692 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1764 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

dhecht commented on Source/JavaScriptCore/dfg/DFGOSRAvailabilityAnalysisPhase.cpp:
```diff
@@ -189,7 +189,8 @@ class OSRAvailabilityAnalysisPhase : public Phase {
 
-                    calculator.executeNode(block->at(nodeIndex));
+                    calculator.executeNode(node);
+                    calculator.validateNode(node);
```
Shouldn't the validateNode come before executeNode? That would be consistent with the existing ASSERTs, right?

dhecht commented on Source/JavaScriptCore/dfg/DFGOSRAvailabilityAnalysisPhase.cpp:
```diff
@@ -189,7 +189,8 @@ class OSRAvailabilityAnalysisPhase : public Phase {
                         }
                     }
```
maybe all this should be moved into validateNode?

dhecht commented on Source/JavaScriptCore/dfg/DFGOSRAvailabilityAnalysisPhase.cpp:
```diff
@@ -139,64 +139,111 @@ class OSRAvailabilityAnalysisPhase : public Phase {
+                Availability argumentCount =
+                    calculator.m_availability.m_locals.operand(VirtualRegister(inlineCallFrame->stackOffset + CallFrameSlot::argumentCountIncludingThis));
+
+                ASSERT_UNUSED(argumentCount, !argumentCount.isDead());
```
Let's make these RELEASE_ASSERT (or DFG_CRASH) since this whole thing is guarded by a runtime flag now, so that this validation happens whenever those flags are enabled.

dhecht commented on Source/JavaScriptCore/dfg/DFGOSRAvailabilityAnalysisPhase.cpp:
```diff
@@ -139,64 +139,111 @@ class OSRAvailabilityAnalysisPhase : public Phase {
+                Availability callee = calculator.m_availability.m_locals.operand(
+                    VirtualRegister(inlineCallFrame->stackOffset + CallFrameSlot::callee));
+
+                ASSERT_UNUSED(callee, !callee.isDead());
```
ditto

dhecht commented on Source/JavaScriptCore/dfg/DFGOSRAvailabilityAnalysisPhase.cpp:
```diff
@@ -139,64 +139,111 @@ class OSRAvailabilityAnalysisPhase : public Phase {
+                Availability argument = calculator.m_availability.m_locals.operand(
+                    VirtualRegister(inlineCallFrame->stackOffset + CallFrame::argumentOffset(i)));
+
+                ASSERT_UNUSED(argument, !argument.isDead());
```
ditto

dhecht reviewed (approved):
r=me with comments

danlliu committed:
DFGOSRAvailabilityAnalysis phase should only assert liveness of arguments during validation pass
https://bugs.webkit.org/show_bug.cgi?id=301773
rdar://158595919

Reviewed by Dan Hecht.

It is incorrect to assert liveness while we are still converging to fixpoint;
we should only assert once we've converged to fixpoint. This fix ensures that
we only check that arguments are live to calls when we have reached fixpoint.

* Source/JavaScriptCore/dfg/DFGOSRAvailabilityAnalysisPhase.cpp:
(JSC::DFG::OSRAvailabilityAnalysisPhase::run):
(JSC::DFG::LocalOSRAvailabilityCalculator::executeNode):
(JSC::DFG::LocalOSRAvailabilityCalculator::validateNode):
* Source/JavaScriptCore/dfg/DFGOSRAvailabilityAnalysisPhase.h:

Canonical link: https://commits.webkit.org/302616@main
